File: python/ccf_rdd.py
# ccf_rdd.py
# CCF algorithm implementation using Spark RDD API.
#
# Algorithm (from: Kardes et al., "CCF: Fast and Scalable Connected
# Component Computation in MapReduce"):
#   Iterates two MapReduce jobs until convergence:
#     - CCF-Iterate : propagates the minimum node ID across neighbors
#     - CCF-Dedup   : removes duplicate pairs to reduce I/O overhead

from pyspark import SparkContext


# CCF-Iterate — Map phase
def _ccf_iterate_map(edge):
    """
    For each pair (a, b), emits both (a, b) and (b, a).
    This makes edges bidirectional so each node sees all its neighbors
    in the subsequent groupByKey reduce step.

    SPARK MECHANIC: By emitting both (a, b) and (b, a), we prepare the RDD 
    for the groupByKey() shuffle phase. This ensures that when Spark partitions 
    the data across workers, every node key will group perfectly with its entire 
    adjacency list, regardless of the original edge direction.
    """
    a, b = edge
    yield (a, b)
    yield (b, a)


# CCF-Iterate — Reduce phase (factory capturing the accumulator)
def _make_reduce_fn(new_pair_acc):
    """
    Returns a reduce function that closes over the Spark accumulator.

    For each node `key` with adjacency list `neighbors`:
      1. Find min_val = min(neighbors)
      2. If min_val < key  → emit (key, min_val)          [key adopts min label]
      3. For every v != min_val → emit (v, min_val)        [propagate label]
         and increment the global NewPair counter.
    If min_val >= key, nothing is emitted (this node is already minimal).

    SPARK MECHANIC: This is a closure that captures the Spark Accumulator. 
    The accumulator object is serialized and sent to worker nodes (executors). 
    Workers can only .add() to it; only the Driver program can read its final value.
    """
    def reduce_fn(kv):
        key, vals = kv
        # vals is a PySpark ResultIterable. We must materialize it to a list 
        # in the executor's memory to find the minimum and iterate over it.
        neighbors = list(vals)
        min_val = min(neighbors)
    
        results = []
    
        # Only act when this node is NOT already the minimum in its neighborhood.
        # If min_val >= key, this node is already the representative → emit nothing.
        if min_val < key:
            # Type-1: this node adopts the smaller label
            results.append((key, min_val))
    
            # Type-2: propagate the smaller label to all OTHER neighbors
            # Each of these is a "new pair" that advances convergence
            for v in neighbors:
                if v != min_val:
                    # SPARK MECHANIC: Workers update the accumulator locally. 
                    # Spark aggregates these back to the driver safely.
                    new_pair_acc.add(1)
                    results.append((v, min_val))
    
        return results


    return reduce_fn


# CCF-Iterate — full job
def ccf_iterate_rdd(pairs_rdd, new_pair_acc):
    """
    One execution of the CCF-Iterate MapReduce job.

    Parameters:
        pairs_rdd    : RDD of (node, component_candidate) integer pairs
        new_pair_acc : Spark Accumulator tracking new cross-pairs emitted

    Returns:
        RDD of (node, new_component_candidate) pairs (before deduplication)
    """
    # Map: make all edges bidirectional to reconstruct adjacency lists
    mapped = pairs_rdd.flatMap(_ccf_iterate_map)

    # Reduce: group neighbors per node, apply label propagation logic
    # SPARK MECHANIC: groupByKey() triggers a massive network shuffle, moving data
    # across the cluster so all values for a single key end up on the same partition.
    # While reduceByKey() is generally more optimized, CCF requires the full adjacency 
    # list to generate cross-pairs, making groupByKey() necessary.
    grouped = mapped.groupByKey()
    return grouped.flatMap(_make_reduce_fn(new_pair_acc))


# CCF-Dedup — full job
def ccf_dedup_rdd(pairs_rdd):
    """
    One execution of the CCF-Dedup MapReduce job.

    CCF-Iterate can emit the same (node, component) pair multiple times
    (from different nodes' reduce steps). This job removes duplicates,
    reducing I/O and memory pressure for the next iteration.
    """
    # SPARK MECHANIC: distinct() triggers another wide transformation (shuffle)
    # under the hood (map -> reduceByKey). This deduplication is critical to prevent 
    # an exponential explosion of data partitions in the next CCF-Iterate step.
    return pairs_rdd.distinct()


# Full CCF loop
def run_ccf_rdd(sc, edges_rdd, verbose=True):
    """
    Runs the full CCF algorithm on the given edge list using the RDD API.

    Iterates CCF-Iterate → CCF-Dedup until the NewPair counter reaches 0,
    signalling that all connected components have been found.

    Parameters:
        sc        : SparkContext
        edges_rdd : RDD of (u, v) undirected edge tuples (integers)
        verbose   : if True, prints per-iteration diagnostics

    Returns:
        result_rdd  : RDD of (node, component_id) pairs.
                      Nodes absent from this RDD are their own component minimum.
        n_iters     : number of iterations until convergence
    """
    pairs = edges_rdd
    iteration = 0

    while True:
        iteration += 1

        # Fresh accumulator for this iteration's NewPair counter
        new_pair_acc = sc.accumulator(0)

        # CCF-Iterate
        iterated = ccf_iterate_rdd(pairs, new_pair_acc)

        # CCF-Dedup
        pairs = ccf_dedup_rdd(iterated)

        # Cache to avoid recomputation; count() triggers lazy evaluation
        # and flushes the accumulator to the driver.
        # SPARK MECHANIC: LINEAGE AND LAZY EVALUATION (CRITICAL)
        # We use .cache() to store the RDD in memory, breaking the recomputation chain.
        # We use .count() as an ACTION. Because Spark is lazy, the map/reduce 
        # transformations above haven't actually run yet. .count() forces the DAG 
        # execution, materializing the RDD and flushing the accumulator to the driver.
        pairs.cache()
        pair_count = pairs.count()

        if verbose:
            print(f"  [RDD] Iter {iteration:>2d} | pairs={pair_count:>8,} "
                  f"| new_cross_pairs={new_pair_acc.value:>8,}")

        # If accumulator is 0, convergence is reached.
        if new_pair_acc.value == 0:
            break

    return pairs, iteration

File: scala/cff-spark/src/main/scala/CCFRdd.scala
// CCFRdd.scala
// CCF algorithm implementation using Spark RDD API.
//
// Mirrors ccf_rdd.py exactly:
//   - CCF-Iterate: flatMap (bidirectional) → groupByKey → flatMap (reduce logic)
//   - CCF-Dedup  : distinct()
//   - Convergence: LongAccumulator (NewPair counter), checked after count()

import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import org.apache.spark.util.LongAccumulator
import scala.collection.mutable.ListBuffer

object CCFRdd {

  // CCF-Iterate

  /**
   * Map phase: for each edge (a, b) emit both (a,b) and (b,a) so every node
   * sees all its neighbors when grouped by key in the reduce phase.
   */
  private def ccfIterateMap(pair: (Int, Int)): Seq[(Int, Int)] = {
    // SPARK MECHANIC: FlatMapping to bidirectional edges is a Narrow Transformation.
    // We emit both directions so that the subsequent groupByKey() shuffle guarantees 
    // all neighbors for a single node are routed to the exact same partition.
    val (a, b) = pair
    Seq((a, b), (b, a))
  }

  /**
   * Reduce phase: for each (key, neighbors):
   *   1. Find minVal = min(neighbors)
   *   2. If minVal < key  → emit (key, minVal)         [type-1: adopt label]
   *      For each v ≠ minVal → emit (v, minVal)         [type-2: propagate]
   *                         + increment NewPair counter
   *   3. If minVal ≥ key  → emit nothing (node is already the minimum)
   */
  private def ccfIterateReduce(
      key: Int,
      neighborsIter: Iterable[Int],
      newPairAcc: LongAccumulator
  ): Seq[(Int, Int)] = {
    // SPARK MECHANIC: Materializing the Iterable to a List happens entirely within 
    // the Executor's JVM heap. For extremely dense nodes (super-nodes), this could 
    // cause GC (Garbage Collection) pressure, which is a known limitation of the RDD API.
    val neighbors = neighborsIter.toList
    val minVal    = neighbors.min
    val results   = ListBuffer[(Int, Int)]()

    if (minVal < key) {
      // Type-1: this node adopts the smaller label
      results += ((key, minVal))

      // Type-2: propagate smaller label to all other neighbors
      neighbors.foreach { v =>
        if (v != minVal) {
          // SPARK MECHANIC: The LongAccumulator is serialized and shipped to executors.
          // Executors perform thread-safe local updates (.add). The driver aggregates 
          // these partial sums only when an Action forces DAG execution.
          newPairAcc.add(1L)
          results += ((v, minVal))
        }
      }
    }
    // If minVal >= key: this node is already the group minimum → nothing emitted

    results.toSeq
  }

  /**
   * One full CCF-Iterate MapReduce job.
   *
   * @param pairsRdd   RDD of (node, componentCandidate) pairs
   * @param newPairAcc Spark LongAccumulator tracking cross-pairs emitted
   * @return RDD of new (node, componentCandidate) pairs before deduplication
   */
  def ccfIterateRdd(pairsRdd: RDD[(Int, Int)], newPairAcc: LongAccumulator): RDD[(Int, Int)] = {
    pairsRdd
      .flatMap(ccfIterateMap)                                        // Map phase
      // SPARK MECHANIC: groupByKey() triggers a Wide Transformation (Network Shuffle).
      // Data is serialized across the cluster. We must use groupByKey over reduceByKey 
      // because CCF requires examining the *entire* adjacency list to emit Type-2 cross-pairs.
      .groupByKey()                                                  // Shuffle
      .flatMap { case (key, vals) =>
        ccfIterateReduce(key, vals, newPairAcc)                      // Reduce phase
      }
  }

  // CCF-Dedup

  /**
   * Removes duplicate (node, componentID) pairs emitted by CCF-Iterate.
   * Reduces I/O and memory pressure for the next iteration.
   */
  def ccfDedupRdd(pairsRdd: RDD[(Int, Int)]): RDD[(Int, Int)] = {
    // SPARK MECHANIC: distinct() is another Wide Transformation (under the hood it is 
    // map(x => (x, null)).reduceByKey((x, y) => x).map(_._1)). Deduplication reduces 
    // shuffle I/O volume for the next iteration.
      pairsRdd.distinct()
  }

  // Full CCF loop

  /**
   * Runs the full CCF algorithm using the RDD API.
   *
   * Iterates CCF-Iterate → CCF-Dedup until the NewPair counter reaches 0,
   * meaning no node has a neighbor with a smaller ID than its current label.
   *
   * The LongAccumulator is reset at each iteration via a fresh instance.
   * count() is called after caching to flush the accumulator to the driver
   * before reading its value — this is the correct Spark accumulator pattern.
   *
   * @param sc        SparkContext
   * @param edgesRdd  RDD of (u, v) undirected edge pairs
   * @param verbose   Print per-iteration diagnostics if true
   * @return (resultRdd, numberOfIterations)
   */
  def runCCFRdd(
      sc: SparkContext,
      edgesRdd: RDD[(Int, Int)],
      verbose: Boolean = true
  ): (RDD[(Int, Int)], Int) = {

    var pairs     = edgesRdd
    var iteration = 0
    var converged = false

    while (!converged) {
      iteration += 1

      // Fresh accumulator per iteration — avoids accumulating across iterations
      val newPairAcc = sc.longAccumulator(s"NewPairCounter_iter$iteration")

      // CCF-Iterate 
      val iterated = ccfIterateRdd(pairs, newPairAcc)

      // CCF-Dedup
      pairs = ccfDedupRdd(iterated)

      // Cache then count() — forces Spark to execute the DAG and flush
      // the accumulator value back to the driver before we read it.
      // SPARK MECHANIC: LAZY EVALUATION AND LINEAGE (CRITICAL)
      // .cache() stores the computed RDD in executor memory, breaking the recomputation 
      // of the lineage tree on subsequent loops.
      // .count() is the ACTION that forces the Spark Engine to physically execute the 
      // MapReduce DAG, flushing the accumulator values safely to the driver.
      pairs.cache()
      val pairCount = pairs.count()

      if (verbose)
        println(f"  [RDD] Iter $iteration%3d | pairs=$pairCount%8d | new_cross_pairs=${newPairAcc.value}%8d")

      // Convergence: no type-2 pairs were generated this iteration
      if (newPairAcc.value == 0L) converged = true
    }

    (pairs, iteration)
  }
}
